test/Microsoft.ML.Tests/Scenarios/BinaryClassification.cs

```diff
@@ -0,0 +1,92 @@
+﻿using Microsoft.ML.Data;
+using Microsoft.ML.Models;
+using Microsoft.ML.Runtime.Api;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Trainers;
+using System.IO;
+using Xunit;
+
+namespace Microsoft.ML.Scenarios
+{
+    public partial class ScenariosTests
+    {
+        public class BreastCancerData
+        {
+            [Column(ordinal: "0")]
+            public float Label;
+            [Column(ordinal: "1-9")]
+            [VectorType(9)]
+            public float[] Features;
+        }
+
+        public class BreastCancerPrediction
+        {
+            [ColumnName("PredictedLabel")]
+            public DvBool Cancerous;
+        }
+
+        [Fact]
+        public void SaveModelToOnnxTest()
+        {
+            string dataPath = GetDataPath(@"breast-cancer.txt");
+            var pipeline = new LearningPipeline();
+
+            pipeline.Add(new Data.TextLoader(dataPath)
+            {
+                Arguments = new TextLoaderArguments
+                {
+                    Separator = new[] { '\t' },
+                    HasHeader = true,
+                    Column = new[]
+                    {
+                        new TextLoaderColumn()
+                        {
+                            Name = "Label",
+                            Source = new [] { new TextLoaderRange(0) },
+                            Type = DataKind.Num
+                        },
+
+                        new TextLoaderColumn()
+                        {
+                            Name = "Features",
+                            Source = new [] { new TextLoaderRange(1, 9) },
+                            Type = DataKind.Num
+                        }
+                    }
+                }
+            });
+
+            pipeline.Add(new FastTreeBinaryClassifier() { NumLeaves = 5, NumTrees = 5, MinDocumentsInLeafs = 2 });
+            
+            PredictionModel<BreastCancerData, BreastCancerPrediction> model = pipeline.Train<BreastCancerData, BreastCancerPrediction>();
+            
+            var modelOutpath = GetOutputPath(Path.Combine("..", "Common", 
+                "Scenario", "BinaryClassification", "BreastCancer"), "SaveModelToOnnxTest.zip");
+
+            DeleteOutputPath(modelOutpath);
+
+            var onnxPath = GetOutputPath(Path.Combine("..", "Common",
+                "Scenario", "BinaryClassification", "BreastCancer"), "SaveModelToOnnxTest.pb");
+
+            DeleteOutputPath(onnxPath);
+
+            var onnxAsJsonPath = GetOutputPath(Path.Combine("..", "Common",
+                "Scenario", "BinaryClassification", "BreastCancer"), "SaveModelToOnnxTest.json");
+
+            DeleteOutputPath(onnxAsJsonPath);
+
+            model.WriteAsync(modelOutpath);
```

Comment:
>model.WriteAsync(modelOutpath); [](start = 12, length = 31)

Where are you awaiting the results of this asynchronous operation?

Reply:
This function does not return a Task object that you can wait on. 

---
In reply to: [191496578](https://github.com/dotnet/machinelearning/pull/248#discussion_r191496578) [](ancestors = 191496578)

Reply:
Assuredly this method does return a `Task`...

[https://github.com/dotnet/machinelearning/blob/9f3076df0d7a7f3ebccc53e0602fa30b90a897fc/src/Microsoft.ML/PredictionModel.cs#L149](https://github.com/dotnet/machinelearning/blob/9f3076df0d7a7f3ebccc53e0602fa30b90a897fc/src/Microsoft.ML/PredictionModel.cs#L149)

You need to wait on it. If you don't you have a race condition. If the task does not complete before you invoked the command below, you are screwed.

Broader point: if hypothetically that method *didn't* return a `Task`, or something else you could wait on, this would actually be a serious issue that would make model writing in any reliable fashion impossible, since there would necessarily be a race condition. Even if this wasn't wrong, I'm a little surprised to see that we're somewhat blase about this.

---
In reply to: [191535338](https://github.com/dotnet/machinelearning/pull/248#discussion_r191535338) [](ancestors = 191535338,191496578)

Reply:
I stand corrected.

---
In reply to: [191564625](https://github.com/dotnet/machinelearning/pull/248#discussion_r191564625) [](ancestors = 191564625,191535338,191496578)